### PR TITLE
Complete Assignment 4

### PR DESCRIPTION
**Pull request by osalih1**

### CODE REVIEW

File: mini-projects/00-jq-rs/Cargo.toml [hunk 1]
```diff
@@ -4,3 +4,8 @@ version = "0.1.0"
 edition = "2021"
 
 [dependencies]
+serde = "1.0"
+serde_json = "1.0"
+regex = "1.5"
+clap = { version = "3.0", features = ["derive"] }
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/Cargo.toml:10]: Using the `preserve_order` would have allowed for preserving the original order when `sort_keys` isn't specified.

File: mini-projects/00-jq-rs/src/main.rs [hunk 2]
```diff
@@ -1,3 +1,131 @@
...
+        .get_matches();
+
+    let color_output = matches.is_present("color-output");
+    let monochrome_output = matches.is_present("monochrome-output");
+    let compact_output = matches.is_present("compact-output");
+    let _sort_keys = matches.is_present("sort-keys");
+    let indent = matches.value_of("indent").unwrap().parse::<u8>().unwrap();
+
+    // Check for conflicting arguments
+    if color_output && monochrome_output {
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/main.rs:70]: By the way, there is also a `conflicts_with` option in clap that can automate this check for you.

File: mini-projects/00-jq-rs/src/filters.rs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    } else {
+        Value::Null
+    }
+}
+
+pub fn add(json: &Value) -> Value {
+    if let Some(array) = json.as_array() {
+        let sum = array
+            .iter()
+            .fold(Value::Null, |acc, item| match (acc, item) {
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/filters.rs:44]: Good thinking to use functional programming here, but the problem is because the accumulator starts as a `Null`, we can only ever enter the third match case since the other two require a `Number` or `String` accumulator and it never gets set as such.

Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/filters.rs:44]: Good thinking to use functional programming here, but the problem is because the accumulator starts as a `Null`, we can only ever enter the third match case since the other two require a `Number` or `String` accumulator and it never gets set as such.

File: mini-projects/00-jq-rs/src/filters.rs
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/filters.rs:1]: Consider using traits/structs/enums to clean this up and introduce shared interfaces that also help your expandability

File: mini-projects/00-jq-rs/src/filters.rs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                }
+            }
+            Value::Array(arr.clone())
+        }
+        _ => Value::Null,
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/filters.rs:90]: I'd like to see more breadth in your test cases - in particular, only a single positive case is being nested; more could be done to test potential failures/edge cases

**Review by amit-kumarh — COMMENTED** [review]
# Rubric For Mini Project 0: JQ

## Submission

### Canvas

1: The link submitted on Canvas is for the pull request containing the
assignment submission.

0: The link submitted on Canvas is not for the pull request containing the
assignment submission.

### Git Branch

1: The branch includes only changes for this assignment.

0: The branch includes commits or changes outside this assignment.

### Git Commit Message

1: All commit messages for the branch are reasonably clear.

0: One or more commit messages for the branch are reasonably clear.

### GitHub Pull Request

1: The pull request is made to the `main` branch of the student's fork of the
course repository.

0: The pull request is made to a branch other than `main` or to a repo that is
not the student's fork of the course repository.

### Points

**Submission Score: 4/4**

## Features
### Command Line Interface
For each of the required CLI arguments `filter-string`, `path-to-json`, `color-output`, `monochrome-output`, `sort-keys`, `indent`, and `compact-output`

0.5: The argument can be correctly parsed and consumed by the rest of the package.

0: The argument is not correctly parsed and propagated to the rest of the package.

Additionally, the command line parsing correctly raises errors to the user when conflicting arguments are inputted.

1.5: Errors are properly determined and bubbled up to the user.

0: Errors are not propagated back to the user

**Comments**

Nice job here - just missing the one edge case of checking that `indent` doesn't conflict with `compact-output` (-0.5)

### Supported Filters
For each of the required filters `identity`, `object-identifier index`, `array index`, `array slice`, `pipe`, and `array iterator`

2: The filter is correctly implemented and properly parses all possible JSON inputs.

1: Minor errors exist in the implement of the filter.

0: Major errors exist in the implementation, or the filter was omitted from the package.

Additionally, comprehensive unit tests are required for each feature:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the filter, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the filter.

0: Tests are missing, or lack significant assessment of the filter's performance.

**Comments**

Identity: Implemented
Object Identifier: The filter is implemented, but not parsed (worked after I added some parsing code) (-1), lacking good tests (-1.5)
Array Index: Implemented, but with incorrect syntax (-1), lacking good tests (-1.5)
Array Slice: Implemented, but with incorrect syntax (-1), lacking good tests (-1.5)
Pipe: Implemented, no tests (-2)
Array Iterator: Not implemented correctly (-2), lacking good tests (-1.5)

For all unit tests - you've got some, but you don't test the parsing logic in `main.rs` at all, and additionally are only testing a single positive case - better tests would check for potential failure modes and edge cases as well. Each (-1.5) is roughly -1 for not testing parsing and -0.5 for lack of breadth on filter tests.

### Built-In Functions
For each of the required built-in functions `add`, `length`, `del`

2: The built-in function is correctly implemented and properly processes all possible JSON inputs.

1: Minor errors exist in the implementation of the operator.

0: Major errors exist in the implementation, or the function was omitted from the package.

Additionally, comprehensive unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the function, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the function.

0: Tests are missing, or lack significant assessment of the function's performance.

**Comments**
Add: Implementation error (see line level comment) (-2), there is one positive unit test but it doesn't pass (-2)
Length: Implemented, lacking good tests (-1.5)
Del: Implemented, but with incorrect syntax. Additionally, doesn't allow for slice deletion (-1). Lacking good tests (-1.5)

Same comments on test cases here. Additionally, the `del` tests only tests object deletion and not array deletion.

### Formatted Printing
For each of the required formats, `colored printing`, `monochrome printing`, `sorted keys`, `indents`, and `compact output`

2: The formatter is correctly implemented and properly displays any JSON object..

1: Minor errors exist in the implementation of the formatter.

0: Major errors exist in the implementation, or the formatter was omitted from the package.

Additionally, unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to ensure that the formatter can correctly handle all expected inputs.

1: Unit tests are present, but lack significant breadth of cases to fully exercise the formatter.

0: Tests are missing, or lack significant assessment of the formatter.

**Comments**
Colored Printing: Attempt made, but major errors (there is some color displayed, but formatting is not correct) (-1). Failing test cases (-2)

Monochrome Printing: Implemented, passing test case.

Sorted Keys: Implemented (see line level comment about `preserve_order`). Test case present.

Indents: Attempt made, but with some errors (indents not even, there's an offset from the correct number of spaces, only operates one level in) (-1). Tests failing (-2)

Compact Output: Implemented, one test case failing due to key order (-1)

### Points
**Features Score: 32.5/61**

## Code Quality

### Design

#### File/Module organization

2: The program is appropriately split into reasonable files/modules

1: The program is somewhat organized, but there's room for improvement (splits aren't idiomatic, or some modules are still too large, etc.)

0: The program is not split at all

**Comments**
You've done some splitting between the formatting code and the filters and it's a nice touch to separate out some of your code into a separate library, but I'd like to see some more here - in particular, parsing code should probably not be living in  `main.rs`. (-1)

#### Scalability

5: The code is architected with scalability in mind. If more features needed to be added to the filtering/printing/CLI, they could easily be integrated.

3: The code was architected with some scalaibility in mind. However, significant refactoring would be needed in order to expand the feature set of this tool. (Possible symptoms include not modularizing enough/not effectively using traits + generics).

0: The code is not scalable. The entire package would need to be refactored in order to expand its features.

**Comments**
There's some room for improvement here as well. In particular, adding a new filter to this system would require wading through the parsing tree in `main.rs` and figuring out where to hook in, and I'd like to see some additional data structures to keep your filters more organized and expandable (like traits, structs and enums). (-3)

#### General Cleanliness

4: The code is effectively documented, especially for sources of user-facing errors. Additionally, function signatures are clear and the code is overall very readable.

2: Attempts at verbose documentation were made, but certain sources of error are unclear OR poor function signatures are used and readers would have slight difficulties while scanning the code.

0: Documentation is severely lacking in several palces and the code is not written in a readable way.

**Comments**
There are little/no comments/documentation present. Function signatures and variable names are readable enough, and the code is not terribly convoluted, but I'd like to see some more here as well. (-3)

### Lint/Format

#### RustFmt

2: RustFmt produces no warnings when run on the changed files

1: RustFmt produces a few minor warnings when run on the changed files

0: RustFmt produces many warnings when run on the changed files

#### Clippy

2: Clippy produces no warnings when run on the changed files

1: Clippy produces a few minor warnings when run on the changed files

0: Clippy produces many warnings when run on the changed files

(-1) One dead code clippy warning.

### Points
**Code Quality Score: 7/15**

## Summary
Add up the point totals here

See line level comments for additional details.

**Total Score: 43.5/80**
